Add file size limits to patient uploader

emmastephenson opened the pull request:
# FRONTEND PULL REQUEST

## Related Issue

- In the patient bulk upload bug bash this morning, we discovered that there are no file size limits on the patient bulk uploader. This PR adds file size and row number limits (50MB and 10,000 rows, respectively).

## Changes Proposed

- Prevent users from uploading files greater than 50MB or 10,000 rows

## Additional Information

Error message for 10,001 rows:
![Screen Shot 2023-01-11 at 3 00 34 PM](https://user-images.githubusercontent.com/80282552/211936496-71810b9f-2f20-4075-bbad-5d743e4a2627.png)

Error message for file over 50MB:
![Screen Shot 2023-01-11 at 3 04 32 PM](https://user-images.githubusercontent.com/80282552/211937051-3df4fdc3-f021-4c9c-8499-75d2ac712fbb.png)

## Testing

Use the following files for testing:
[CSV over 10,000 rows](https://drive.google.com/file/d/13c0W4y-6SIT27Z0887yYyma91tktnLc6/view?usp=share_link)
[CSV over 50MB](https://drive.google.com/file/d/1-WW7LQB-T7QsQQaiqY7SXmlveuKvpA_9/view?usp=share_link)

## Checklist for Author and Reviewer
### Accessibility
- [x] Any large changes have been run through Deque manual testing
- [x] All changes have run through the Deque automated testing

### Design
- [ ] Any UI/UX changes have a designer as a reviewer, and changes have been approved
- [ ] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke-tested by both the engineering and design teams

### Content
- [x] Any content changes have been approved by content team (Content provided by @lauraykerr)

### Security
- [x] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.) -> Discussed size limits with @rin-skylight

emyl3 reviewed (changes requested):
one thing I noticed was previous errors do not clear when uploading a file that has too many rows or is over 50MB.
To repro:
- Upload a csv file with errors (errors that are not too many rows or over 50MB)
- See the error messages table
- Upload a csv that is too many rows or over 50MB and you will see the following
<img width="731" alt="Screen Shot 2023-01-11 at 20 38 59" src="https://user-images.githubusercontent.com/20211771/211963257-6e9a7de9-05e6-4242-b63c-b9db8ab5ab48.png">

lauraykerr reviewed (changes requested):
A few additional things I noticed: 

**1. Error pops up before you click the upload button**  
With the other errors, the user clicks the button first and then the error message comes up. I think this behavior is ok and the immediate feedback makes sense, but would suggest changing the header on the message to reflect the difference. In each case, can we change it to: `Error: File too large`?


**2. Previous file doesn't clear after error message**

Steps to reproduce: 
1. Upload test file 
2. See the same file in the upload box
3. Upload new file by clicking "choose file" or refreshing page

Expected behavior: When the user gets other errors, the upload box clears and you have the opportunity to drag a new file into the box. 


**3. Uploading edited file doesn't resolve the error**

Steps to reproduce:
1. Upload test file "fake_patients_over_50mb.csv"
2. Get error message
3. Open file in Excel and remove all rows but first 100
5. Save file
6. Confirm file size in finder
7. Upload edited and saved file 
8. See same error message

Expected behavior: Editing and saving file should resolve the error.

emmastephenson commented:
@emyl3 and @lauraykerr, thank you both for the through reviews! 🙌  I believe I've addressed all your feedback. Some notable changes are:
1. The size validation now occurs _after_ the user hits the "upload" button. This was the only way I could find to get around the third issue you noticed, Laura - the HTML `<input>` element itself doesn't register re-uploading the same file with different contents, so the function where I was originally checking size wasn't being triggered at all. This also fixed the issue of errors not clearing properly.
2. The error header is now customizable, not just the error body. (@lauraykerr let me know if this is ok, since the functionality with the button changed - if you prefer we can just keep the same error message for all issues)
3. These changes fixed our weird test failure from yesterday @emyl3! I did add some new tests for the size checks.

emyl3 commented on frontend/src/config/constants.ts:
```diff
@@ -526,3 +526,6 @@ export const organizationCreationSteps = [
   { label: "Experian identity verification", value: "2", order: 2 },
 ];
+
+export const MAX_CSV_UPLOAD_BYTES = 50 * 1000 * 1000;
```
ooo thank you for moving these here!

emyl3 commented on frontend/src/app/testResults/uploads/Uploads.tsx:
```diff
@@ -52,11 +54,14 @@ const Uploads = () => {
-          minimumFractionDigits: 2,
-        });
+      if (currentFile.size > MAX_CSV_UPLOAD_BYTES) {
+        const maxKBytes = (MAX_CSV_UPLOAD_BYTES / 1024).toLocaleString(
```
if you upload a file over 50MB then the file does not clear unlike other errrors 🤔 From this diff, I'm not seeing why that's happening.
<img width="705" alt="Screen Shot 2023-01-13 at 09 11 51" src="https://user-images.githubusercontent.com/20211771/212355876-efb49b3e-9190-475b-b38b-09751bb9dc46.png">

emmastephenson replied:
Resolved offline! We needed to move up the `await file.text()` call to force the file to load, which we _think_ triggered an onChange for the `input` element.

emyl3 reviewed (commented):
So glad you didn't have to make any of the hacky changes I suggested to get those older frontend tests to pass. 😆 I'm noticing there are new frontend tests failure in CI and that is reproduce-able locally. Let me know if you need an extra pair of eyes on this!

I confirmed the issue I noted previously is no longer happening. Thank you for addressing that!

emmastephenson commented:
@emyl3 and @lauraykerr, ready for re-review (again!) 😄

lauraykerr reviewed (commented):
@emmastephenson Most of the issues from before appear resolved, and I'm the customized header looks great. I'm still getting some weird behavior, though, as the "too many rows" error is occasionally (and seemingly inconsistently) popping up even when I fix the CSV. 

Steps to reproduce: 

1. Upload `fake_patients_too_many_rows`
2. Get the "too many rows" error 
3. Delete rows to row 10,000 and save
5. Upload file again
6. See error 
7. Delete rows to 9,999 and save
8. Upload file again
9. Maybe see error?

Once when I tested this, I deleted to 9,999 and it worked. Another time it didn't work until I deleted a few more rows. One time I added rows to a file that had worked with fewer rows and it worked on 9,999 but not 10,000, then when I tried it again with 9,999 it didn't work.

emyl3 reviewed (approved):
Worked for me as intended! I tried uploading a 10k+ row csv, then changed it to 9999 rows and then waited for around 40ish mins and saw my patients.

nice work emma!!!

emmastephenson commented:
@lauraykerr thanks for the feedback! I'm not able to repro the failed upload with 9,999 rows, but I do know why the 10,000 row file failed - the header counts as a row, so that file technically has 10,001 rows. I'm a little hesitant to change that only because it's also the limit on the results uploader, and I _think_ that might be a hard limit on the ReportStream side. @zdeveloper can you confirm?

lauraykerr commented:
@emmastephenson Took a [screen recording of getting the error](https://drive.google.com/file/d/1IvKCIl49upHEmQEK1n65K__8VCUMi9tv/view?usp=share_link) on 9,999 rows in case that reveals anything (apologies for my lack of keyboard shortcuts). Happy to also pair at some point to walk through it. 

The 10,000 rows makes sense. In that case, though, can I make a request for a slight edit on the error messages? Sorry to drag this out!

- “[File_name]” has too many rows for SimpleReport to process. Please limit each upload to less than XX,XXX rows.
- “[File_name]” is too large for SimpleReport to process. Please limit each upload to less than 50 MB.

emmastephenson commented:
@lauraykerr thanks! I changed the error messages to the new copy. If you're ok with it, I'm going to punt on the 9,999 rows issue until we see if it's reproducible in production, since I'm still not able to reproduce.

lauraykerr commented:
@emmastephenson Thanks, and sounds good to me!

johanna-skylight commented on frontend/src/app/patients/UploadPatients.tsx:
```diff
@@ -50,25 +59,42 @@ const UploadPatients = ({ isAdmin }: { isAdmin: boolean }) => {
+  ) {
+    setErrorMessageHeader(header);
+    setErrorMessageText(body);
+    setIncludeGuideInErrorMessage(includeGuide);
```
Do these three states really need to be separate states?. Could a single state with a more specific data structure work as well? For example:

type ErrorMessage = {
text: string,
header?:string,
includeGuide:boolean
}

This will allow to save some renders

emmastephenson replied:
Thank you for the suggestion! I like this much better, changed to a single `errorMessage` state.

johanna-skylight commented on frontend/src/app/patients/UploadPatients.tsx:
```diff
@@ -281,14 +336,19 @@ const UploadPatients = ({ isAdmin }: { isAdmin: boolean }) => {
+                                </a>{" "}
+                                for details about accepted values.
+                              </div>
+                            )}
```
A nice to have but non blocker would be taking the DOM of the error alert out of the general render and to a function

emmastephenson replied:
Done 👍

johanna-skylight commented on frontend/src/app/patients/UploadPatients.tsx:
```diff
@@ -281,14 +336,19 @@ const UploadPatients = ({ isAdmin }: { isAdmin: boolean }) => {
-                            for details about accepted values.
+                            {errorMessageText}
+                            {includeGuideInErrorMessage && (
+                              <div>
```
div element needs to be switched to a span because is within a paragraph element. To make it behave like a div just add the class display-block
`<span class="display-block">`

emmastephenson replied:
Thanks for catching this! Sharp eye 👀

johanna-skylight reviewed (commented):
Main functionality looks good! Left some comments on small things that can be improved

emyl3 reviewed (approved):
New copy updates look good to me from my last review! Let me know if you want to address johanna's comments in this PR (happy to re-review afterwards! :)) or in a different PR.

lauraykerr reviewed (dismissed):
LGTM! Thanks for the copy changes @emmastephenson

emyl3 reviewed (approved):
Thank you for all your work on this!!! 👍
